source/common/network/io_socket_handle_impl.cc

```diff
@@ -416,8 +457,23 @@ Api::IoCallUint64Result IoSocketHandleImpl::recvmmsg(RawSliceArrays& slices, uin
 
     output.msg_[i].msg_len_ = mmsg_hdr[i].msg_len;
     // Get local and peer addresses for each packet.
-    output.msg_[i].peer_address_ = Address::addressFromSockAddrOrDie(
-        raw_addresses[i], hdr.msg_namelen, fd_, socket_v6only_ || !udp_read_normalize_addresses_);
+    if (recent_received_addresses_ != nullptr) {
+      quic::QuicSocketAddress peer_quic_address(raw_addresses[i]);
+      auto it = recent_received_addresses_->Lookup(peer_quic_address);
+      if (it == recent_received_addresses_->end()) {
+        Address::InstanceConstSharedPtr new_address =
+            Address::addressFromSockAddrOrDie(raw_addresses[i], hdr.msg_namelen, fd_,
+                                              socket_v6only_ || !udp_read_normalize_addresses_);
+        output.msg_[i].peer_address_ = new_address;
+        recent_received_addresses_->Insert(
+            peer_quic_address, std::make_unique<Address::InstanceConstSharedPtr>(new_address));
+      } else {
+        output.msg_[i].peer_address_ = *it->second;
+      }
+    } else {
+      output.msg_[i].peer_address_ = Address::addressFromSockAddrOrDie(
+          raw_addresses[i], hdr.msg_namelen, fd_, socket_v6only_ || !udp_read_normalize_addresses_);
+    }
```

Comment:
Hm. It seems like we have 4 places where we're doing basically the same thing of looking up an address in the cache and possible creating a new one. Can we share this code somehow, perhaps with a helper method?

Reply:
Moved them into a member function.